File: packages/face-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/facedetection/FaceDetectionPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.facedetection;

import android.graphics.PointF;
import android.graphics.Rect;
import android.net.Uri;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceContour;
import com.google.mlkit.vision.face.FaceDetectorOptions;
import com.google.mlkit.vision.face.FaceLandmark;
import io.capawesome.capacitorjs.plugins.mlkit.facedetection.classes.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.facedetection.classes.ProcessImageResult;
import java.util.List;

@CapacitorPlugin(name = "FaceDetection")
public class FaceDetectionPlugin extends Plugin {

    public static final String TAG = "FaceDetection";
    public static final String ERROR_PROCESS_IMAGE_CANCELED = "processImage canceled.";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_LOAD_IMAGE_FAILED = "image could not be loaded.";

    private FaceDetection implementation;

    @Override
    public void load() {
        try {
            implementation = new FaceDetection(this);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void processImage(PluginCall call) {
        try {
            String path = call.getString("path", null);
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            Integer performanceMode = call.getInt("performanceMode", FaceDetectorOptions.PERFORMANCE_MODE_FAST);
            Integer landmarkMode = call.getInt("landmarkMode", FaceDetectorOptions.LANDMARK_MODE_NONE);
            Integer contourMode = call.getInt("contourMode", FaceDetectorOptions.CONTOUR_MODE_NONE);
            Integer classificationMode = call.getInt("classificationMode", FaceDetectorOptions.CLASSIFICATION_MODE_NONE);
            Float minFaceSize = call.getFloat("minFaceSize", 0.1f);
            Boolean enableTracking = call.getBoolean("enableTracking", false);

            InputImage image = implementation.createInputImageFromFilePath(path);
            if (image == null) {
                call.reject(ERROR_LOAD_IMAGE_FAILED);
                return;
            }
            ProcessImageOptions options = new ProcessImageOptions(
                image,
                performanceMode,
                landmarkMode,
                contourMode,
                classificationMode,
                minFaceSize,
                enableTracking
            );

            implementation.processImage(
                options,
                new ProcessImageResultCallback() {
                    @Override
                    public void success(ProcessImageResult result) {
                        call.resolve(result.toJSObject());
                    }

                    @Override
                    public void cancel() {
                        call.reject(ERROR_PROCESS_IMAGE_CANCELED);
                    }

                    @Override
                    public void error(Exception exception) {
                        String message = exception.getMessage();
                        Logger.error(TAG, message, exception);
                        call.reject(message);
                    }
                }
            );
        } catch (Exception exception) {
            String message = exception.getMessage();
            Logger.error(TAG, message, exception);
            call.reject(message);
        }
    }
}

File: packages/face-detection/ios/Plugin/FaceDetectionPlugin.swift
import Foundation
import Capacitor
import MLKitVision
import MLKitFaceDetection

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FaceDetectionPlugin)
public class FaceDetectionPlugin: CAPPlugin {
    public let tag = "FaceDetection"
    public let errorPathMissing = "path must be provided."
    public let errorLoadImageFailed = "image could not be loaded."

    private var implementation: FaceDetection?

    override public func load() {
        implementation = FaceDetection(plugin: self)
    }

    @objc func processImage(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        let performanceMode = call.getInt("performanceMode", FaceDetectorPerformanceMode.fast.rawValue)
        let landmarkMode = call.getInt("landmarkMode", FaceDetectorLandmarkMode.none.rawValue)
        let contourMode = call.getInt("contourMode", FaceDetectorContourMode.none.rawValue)
        let classificationMode = call.getInt("classificationMode", FaceDetectorClassificationMode.none.rawValue)
        let minFaceSize = call.getFloat("minFaceSize", 0.1)
        let enableTracking = call.getBool("enableTracking", false)

        guard let visionImage = implementation?.createVisionImageFromFilePath(path) else {
            call.reject(errorLoadImageFailed)
            return
        }

        let options = ProcessImageOptions(visionImage: visionImage, performanceMode: performanceMode, landmarkMode: landmarkMode,
                                          contourMode: contourMode, classificationMode: classificationMode, minFaceSize: minFaceSize, enableTracking: enableTracking)

        implementation?.processImage(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, nil, error)
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }
}
